File: backend/src/utils/haversine.js
/**
 * haversine.js — Standalone, unit-tested Haversine distance calculator.
 *
 * Returns the great-circle distance in kilometres between two lat/lng points.
 * Used in 3 places:
 *   1. POST /api/visits/login   → distance_from_previous_km
 *   2. POST /api/visits/logout  → out_of_radius check vs dealer lat/lng
 *   3. POST /api/attendance/logout → total_distance_km accumulation
 */

const EARTH_RADIUS_KM = 6371;

/**
 * @param {number} lat1 - Latitude of point A (degrees)
 * @param {number} lng1 - Longitude of point A (degrees)
 * @param {number} lat2 - Latitude of point B (degrees)
 * @param {number} lng2 - Longitude of point B (degrees)
 * @returns {number} Distance in kilometres
 */
function haversineKm(lat1, lng1, lat2, lng2) {
  const toRad = (deg) => (deg * Math.PI) / 180;

  const dLat = toRad(lat2 - lat1);
  const dLng = toRad(lng2 - lng1);

  const a =
    Math.sin(dLat / 2) * Math.sin(dLat / 2) +
    Math.cos(toRad(lat1)) *
      Math.cos(toRad(lat2)) *
      Math.sin(dLng / 2) *
      Math.sin(dLng / 2);

  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
  return EARTH_RADIUS_KM * c;
}

/**
 * Returns true if point (lat2, lng2) is within `radiusMeters` of (lat1, lng1).
 * @param {number} lat1
 * @param {number} lng1
 * @param {number} lat2
 * @param {number} lng2
 * @param {number} radiusMeters
 * @returns {boolean}
 */
function isWithinRadius(lat1, lng1, lat2, lng2, radiusMeters) {
  const distKm = haversineKm(lat1, lng1, lat2, lng2);
  return distKm * 1000 <= radiusMeters;
}

module.exports = { haversineKm, isWithinRadius };

File: backend-py/app/utils/haversine.py
"""
haversine.py — ports haversine.js exactly. Great-circle distance in
kilometres. NOTE (see INVENTORY.md §8.5): only used for radius/geofence
checks (login/logout/location-check distance-to-dealer) — NOT for any
distance-travelled totals, which go through Google Routes API with no
fallback (see app/services/google_routes.py).
"""
import math

EARTH_RADIUS_KM = 6371


def haversine_km(lat1: float, lng1: float, lat2: float, lng2: float) -> float:
    to_rad = math.radians
    d_lat = to_rad(lat2 - lat1)
    d_lng = to_rad(lng2 - lng1)

    a = (
        math.sin(d_lat / 2) ** 2
        + math.cos(to_rad(lat1)) * math.cos(to_rad(lat2)) * math.sin(d_lng / 2) ** 2
    )
    c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))
    return EARTH_RADIUS_KM * c


def is_within_radius(lat1: float, lng1: float, lat2: float, lng2: float, radius_meters: float) -> bool:
    return haversine_km(lat1, lng1, lat2, lng2) * 1000 <= radius_meters
